### PR TITLE
[SIL] Allow giving debug names to SILBasicBlocks

### PR DESCRIPTION
**Pull request by ktoso**
Resolves ancient issue rdar://19283513

It is very hard to wrap your head around more complex and dynamically generated basic block graphs without the ability to tag them with some human readable name.

This PR solves this, by allowing SILBasicBlock to have a "debug name". The names are stored in the Module, and are only available in `NDEBUG` builds.

### CODE REVIEW

**Comment by ktoso** [comment]
resolves rdar://19283513

Would you have a look @atrick - is this what you had in mind here? I may be missing some common patterns or tricks how we usually do this in the compiler. Thanks in advance, this has been a lifesaver in my recent work so I'd like to actually include it for good :-)

File: lib/SIL/IR/SILPrinter.cpp [hunk 1]
```diff
@@ -816,6 +816,12 @@ class SILPrinter : public SILInstructionVisitor<SILPrinter> {
       for (auto Id : PredIDs)
         *this << ' ' << Id;
     }
+
+    // If the basic block has a name available, print it as well
+    auto debugName = BB->getDebugName();
+    if (debugName.hasValue()) {
+      *this << " /// " << debugName.getValue();
+    }
```
Comment (ktoso) [inline lib/SIL/IR/SILPrinter.cpp:824]: open to other ideas where to print it; wasn't quite sure how it is printed in IR, I suppose we could make it the same as there?

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: include/swift/SIL/SILModule.h [hunk 2]
```diff
@@ -163,6 +163,8 @@ class SILModule {
       llvm::ilist<SILDifferentiabilityWitness>;
   using CoverageMapCollectionType =
       llvm::MapVector<StringRef, SILCoverageMap *>;
+  using BasicBlockNameMapType =
+      llvm::DenseMap<const SILBasicBlock *, llvm::StringRef>;
```
Comment (atrick) [inline include/swift/SIL/SILModule.h:167]: From the `StringRef` comments
```
  /// This class does not own the string data, it is expected to be used in
  /// situations where the character data resides in some other buffer, whose
  /// lifetime extends past that of the StringRef. For this reason, it is not in
  /// general safe to store a StringRef.
```

I expected the stored type to be a `std::string`. I think if you just change the type in the map everything will just work, and the API's can still accept and vend `StringRef`, which is suitable for printing.

File: lib/SIL/IR/SILPrinter.cpp (hunk 1 above)
Comment (atrick) [reply lib/SIL/IR/SILPrinter.cpp:824]: I would print this the same way we do for demangled names--print a separate comment line *before* the block ID.

As long as the .sil still parses, it's a matter of taste. You can check by piping the output back through swift-frontend or sil-opt.

File: include/swift/SIL/SILModule.h (hunk 2 above)
Comment (ktoso) [reply include/swift/SIL/SILModule.h:167]: Oh, didn't realize that -- some other collections did store StringRef here so I followed that.

Moving to std::string then 👍

File: lib/SIL/IR/SILPrinter.cpp (hunk 1 above)
Comment (ktoso) [reply lib/SIL/IR/SILPrinter.cpp:824]: That sounds good, will do 👍

**Comment by ktoso** [comment]
Addressed the comments, please have a look

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Review by kavon — APPROVED** [review]

**Review by atrick — APPROVED** [review]
LGTM

**Comment by ktoso** [comment]
Sweet, thanks for the advice @atrick !

**Pull request merged**